src/Database/PostgreSQL/LibPQ.hsc

```diff
@@ -2636,33 +2694,42 @@ foreign import ccall        "libpq-fe.h PQunescapeBytea"
                       -> Ptr CSize
                       -> IO (Ptr Word8) -- Actually (IO (Ptr CUChar))
 
-foreign import ccall unsafe "libpq-fe.h PQescapeIdentifier"
+-- Must not be `unsafe` because `PQescapeIdentifier` may call `libpq_gettext()` that may do IO.
```

Comment:
Does it actually block? Aren't translations loaded once into memory?

Reply:
I think they are loaded on demand:

https://github.com/postgres/postgres/blob/70f7da3e6ec865a120f3a3fed6173ce20f7d485a/src/interfaces/libpq/fe-misc.c#L1128-L1153

Here, `bindtextdomain()` sets the directory from which translation files are read, and `dgettext()` returns the translation string, so in between disk IO must be happening in practice.

Postgres puts an `if (!already_bound)` around it, but that is quite the implementation detail; to make `PQescapeIdentifier()` `unsafe`, we'd have to find some guarantee that it is imposible to call this function without some other `safe` function triggering the gettext code path first. This would be quite fragile.